File: ablibrary/src/main/java/cc/ab/base/widget/discretescrollview/adapter/DiscretePageAdapter.java
package cc.ab.base.widget.discretescrollview.adapter;

import android.view.*;
import androidx.recyclerview.widget.RecyclerView;
import cc.ab.base.ext.ViewExtKt;
import cc.ab.base.widget.discretescrollview.holder.DiscreteHolder;
import cc.ab.base.widget.discretescrollview.holder.DiscreteHolderCreator;
import cc.ab.base.widget.discretescrollview.listener.OnItemClickListener;
import java.util.List;
import org.jetbrains.annotations.NotNull;

/**
 * Description:
 *
 * @author: CASE
 * @date: 2019/10/14 12:09
 */
public class DiscretePageAdapter<T> extends RecyclerView.Adapter<DiscreteHolder<T>> {
  private List<T> datas;
  private DiscreteHolderCreator creator;
  private OnItemClickListener onItemClickListener;

  public DiscretePageAdapter(DiscreteHolderCreator creator, List<T> datas) {
    this.creator = creator;
    this.datas = datas;
  }

  @SuppressWarnings("unchecked")
  @NotNull @Override
  public DiscreteHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    int layoutId = creator.getLayoutId();
    View itemView = LayoutInflater.from(parent.getContext()).inflate(layoutId, parent, false);
    return creator.createHolder(itemView);
  }

  @SuppressWarnings("unchecked")
  @Override
  public void onBindViewHolder(DiscreteHolder holder, final int position) {
    holder.updateUI(datas.get(position), position, getItemCount());

    if (onItemClickListener != null) {
      holder.itemView.setOnClickListener(v -> onItemClickListener.onItemClick(position, datas.get(position)));
      ViewExtKt.pressEffectAlpha(holder.itemView, 0.9f);
    } else {
      ViewExtKt.pressEffectDisable(holder.itemView);
    }
  }

  @Override
  public int getItemCount() {
    return datas == null ? 0 : datas.size();
  }

  public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
    this.onItemClickListener = onItemClickListener;
  }
}

File: ablibrary/src/main/java/cc/ab/base/widget/discretescrollview/adapter/DiscretePageAdapter.kt
package cc.ab.base.widget.discretescrollview.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import androidx.viewbinding.ViewBinding
import cc.ab.base.ext.*
import cc.ab.base.ui.item.BaseViewHolder
import cc.ab.base.widget.discretescrollview.listener.OnItemClickListener
import com.dylanc.viewbinding.base.inflateBindingWithGeneric

/**
 * @Description
 * @Author：Khaos
 * @Date：2021/3/16
 * @Time：18:18
 */
abstract class DiscretePageAdapter<T, V : ViewBinding>(
  private val datas: MutableList<T> = mutableListOf(),
  var onItemClickListener: OnItemClickListener<T>? = null,
) : RecyclerView.Adapter<BaseViewHolder<V>>() {

  //<editor-fold defaultstate="collapsed" desc="创建ViewHolder">
  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder<V> {
    return BaseViewHolder(this.inflateBindingWithGeneric(LayoutInflater.from(parent.context), parent, false))
  }
  //</editor-fold>

  //<editor-fold defaultstate="collapsed" desc="绑定数据">
  override fun onBindViewHolder(holder: BaseViewHolder<V>, position: Int) {
    //更新数据
    fillData(datas[position], holder.viewBinding, position, itemCount)
    //点击事件
    if (onItemClickListener != null) {
      holder.itemView.click { onItemClickListener?.onItemClick(position, datas[position]) }
      holder.itemView.pressEffectAlpha(0.9f)
    } else {
      holder.itemView.pressEffectDisable()
    }
  }
  //</editor-fold>

  //<editor-fold defaultstate="collapsed" desc="数据大小">
  override fun getItemCount() = datas.size
  //</editor-fold>

  //<editor-fold defaultstate="collapsed" desc="子类必须重新的方法">
  //加载数据
  abstract fun fillData(data: T, binding: V, position: Int, count: Int)
  //</editor-fold>
}